Source/WebCore/loader/HistoryController.cpp

```diff
@@ -336,49 +347,80 @@ struct HistoryController::FrameToNavigate {
-        // If the navigation was aborted (by the JS called preventDefault() on the navigate event), then
-        // do not do any further navigations.
-        if (abortHandler->wasAborted()) {
-            triggeringFrame.window()->protectedNavigation()->rejectFinishedPromise(tracker);
-            break;
+        // Set the BF cursor before commit, which lets the user quickly click back/forward again.
+        // - plus, it only makes sense for the top level of the operation through the frame tree,
+        // as opposed to happening for some/one of the page commits that might happen soon
+        CheckedRef backForward = page->backForward();
+        RefPtr currentItem = backForward->currentItem(m_frame->frameID());
+        backForward->setProvisionalItem(targetItem);
+
+        // First set the provisional item of any frames that are not actually navigating.
+        // This must be done before trying to navigate the desired frame, because some
+        // navigations can commit immediately (such as about:blank). We must be sure that
+        // all frames have provisional items set before the commit.
+        recursiveSetProvisionalItem(targetItem, currentItem.get(), ForNavigationAPI::Yes);
+
+        for (auto& frameToNavigate : framesToNavigate) {
+            Ref abortHandler = frameToNavigate.frame->window()->protectedNavigation()->registerAbortHandler();
+            frameToNavigate.frame->protectedLoader()->loadItem(frameToNavigate.toItem, frameToNavigate.fromItem.get(), frameLoadType, ShouldTreatAsContinuingLoad::No);
+            // If the navigation was aborted (by the JS called preventDefault() on the navigate event), then
+            // do not do any further navigations.
+            if (abortHandler->wasAborted()) {
+                triggeringFrame->window()->protectedNavigation()->rejectFinishedPromise(tracker.get());
+                break;
+            }
         }
+    };
+
+    goToItemShared(targetItem, WTFMove(finishGoToItem));
+}
+
+void HistoryController::goToItemShared(HistoryItem& targetItem, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_policyItem = &targetItem;
+
+    // Same document navigations must continue synchronously from here.
+    // Therefore asynchronous `should go to history item` checks cannot prevent them,
+    // and are informational only.
+    RefPtr current = currentItem();
+    bool sameDocumentNavigation = current && targetItem.shouldDoSameDocumentNavigationTo(*current);
+
+    if (sameDocumentNavigation || !m_frame->loader().client().supportsAsyncShouldGoToHistoryItem()) {
+        bool result = m_frame->loader().client().shouldGoToHistoryItem(targetItem, sameDocumentNavigation);
+        completionHandler(result);
+        return;
     }
+
+    m_frame->loader().client().shouldGoToHistoryItemAsync(targetItem, [completionHandler = WTFMove(completionHandler)] (bool result) mutable {
```

Comment:
Can we do this?
```
m_frame->loader().client().shouldGoToHistoryItemAsync(targetItem, WTFMove(completionHandler));
```

Reply:
Yes - this weird extra lambda made its way in during an earlier version of the patch, but the need for it has factored out 👍